### PR TITLE
Documentation first steps

### PR DESCRIPTION
**Pull request by DavidAndreev**
Please give me some feedback on the documentation I wrote so far, main file is the my_documentation.rst .If you click view it will nicely render the .rst file. Please check mostly content if its correct.

### CODE REVIEW

File: docs/my_documentation.rst [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    multipass = Multipass()
+    for app in apps:
+        multipass.init_app(app)
+
+The following configuration values exist for Flask-Multipass:
+
+====================================== =========================================
+``MULTIPASS_AUTH_PROVIDERS``           Dictionary of authentication providers
+``MULTIPASS_IDENTITY_PROVIDERS``       Dictionary of identification providers
+``MULTIPASS_PROVIDER_MAP`` `           Mapping of authentication providers to identification providers
```
Comment (mvidalgarcia) [inline docs/my_documentation.rst]: Extra `

File: docs/my_documentation.rst [hunk 2]
```diff
@@ -0,0 +1,365 @@
+Flask-Multipass
+===============
+
+Installation
+------------
+
+Installing Flask-Multipass is very easy using pip::
```
Comment (pferreir) [inline docs/my_documentation.rst]: > Installing Flask-Multipass is very easy. Using pip::

File: docs/my_documentation.rst [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+
+    $ pip install Flask-Multipass
+
+Initialization
+--------------
+
+To start with Multipass, create your Flask application, load the prefered configuration and create the ``Multipass`` object by passing it the application:
```
Comment (pferreir) [inline docs/my_documentation.rst]: "Preferred". It would be nice to pass the whole file through a spellchecker.

File: docs/my_documentation.rst [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+.. code-block:: python
+
+    from flask import Flask
+    from flask_multipass import Multipass
+
+    app = Flask(__name__)
+    app.config.from_pyfile('example.cfg')
+    multipass = Multipass(app)
+
+When you have more then one application or you are using an application factory you can also use ``init_app`` :
```
Comment (pferreir) [inline docs/my_documentation.rst]: > more **than** one

Comment (pferreir) [inline docs/my_documentation.rst]: Actually, I would rephrase it:

> If you happen to have more than one application (or you're using an application factory) ...

File: docs/my_documentation.rst [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+``MULTIPASS_LOGIN_ENDPOINT``           Endpoint linking to login page
+``MULTIPASS_LOGIN_URLS``               List of login URLs 
+``MULTIPASS_SUCCESS_ENDPOINT``         Endpoint linking to default page after successful login
+``MULTIPASS_FAILURE_MESSAGE``          Message to show after unsuccessfull login
+``MULTIPASS_FAILURE_CATEGORY``         Category of message when flashing after unsuccessful login
+``MULTIPASS_ALL_MATCHING_IDENTITIES``  If true, all matching identities are passed after successful authentication
+``MULTIPASS_REQUIRE_IDENTITY``         If true, ``IdentityRetrievalFailed`` is raised when no matching identities are found, othwerwise empty list is passed
+====================================== =========================================
+
+Configuration example can be found here
```
Comment (pferreir) [inline docs/my_documentation.rst]: "A configuration example" or "An example configuration file"

File: docs/my_documentation.rst [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+
+< link to config.example.rst>
+
+Registering providers
+---------------------
+
+You can register new provider using the following commands. Remember that your provider class must be a subclass of either ``AuthProvider`` or ``IdentityProvider``:
```
Comment (pferreir) [inline docs/my_documentation.rst]: > You can register **a** new provider

and "remember" implies that the person has already had contact with the library, which isn't true. I'd use "keep in mind" instead.

Comment (pferreir) [inline docs/my_documentation.rst]: You say "You can register new provider using the following commands." but what you show immediately below are not the commands. I'd rather say something like:

> We'll start by registering a new authentication provider. First of all, let's create the corresponding class, which in this case should inherit from ``AuthProvider`` (it would be ``IdentityProvider`` for an identity provider).

File: docs/my_documentation.rst [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+
+.. code-block:: python
+
+	class FooProvider(AuthProvider):
+		pass
+
+Then you pass this class to ``register_provider`` with second argument beeing the type name of the provider specified in configuration. Remember that type names must be unique. :
```
Comment (pferreir) [inline docs/my_documentation.rst]: > with **the** second argument

But I would rephrase it according to what I wrote before:

>  Then, we'll call ``register_provider``, which will associate a provider class with a unique id that is passed as the second argument:

File: docs/my_documentation.rst [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+
+.. code-block:: python
+
+	app = Flask('test')
+	app.config['MULTIPASS_AUTH_PROVIDERS'] = {
+		'test_auth_provider': {'type': 'test'}
+	}
+	multipass = Multipass(app)
```
Comment (pferreir) [inline docs/my_documentation.rst]: You've already written these lines above, there's no need to repeat them, as people should be following along.

File: docs/my_documentation.rst [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+	multipass.register_provider(FooProvider, 'test')
+	
+To instantiate new providers you need to call ``_create_providers`` with key and base class arguments referring to type of providers you registered:
```
Comment (pferreir) [inline docs/my_documentation.rst]: Do we really need to call this? I don't recall that.

Comment (pferreir) [inline docs/my_documentation.rst]: /cc @ThiefMaster

File: docs/my_documentation.rst [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+or:
+
+.. code-block:: python
+
+	multipass._create_providers('IDENTITY', IdentityProvider))
+	
+Accessing providers
+---------------------
+	
+Providers can be access through ``auth_providers`` property:
```
Comment (pferreir) [inline docs/my_documentation.rst]: > through **the** ``auth_providers`` property:

File: docs/my_documentation.rst [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+.. code-block:: python
+
+	def _create_login_rule(self):
+		"""Creates the login URL rule if necessary"""
+		endpoint = current_app.config['MULTIPASS_LOGIN_ENDPOINT']
+		rules = current_app.config['MULTIPASS_LOGIN_URLS']
+		if rules is None:
+		    return
+		for rule in rules:
+		    current_app.add_url_rule(rule, endpoint, self.process_login, methods=('GET', 'POST'))
```
Comment (pferreir) [inline docs/my_documentation.rst]: I wouldn't really specify this code block here. This is an internal method of multipass and people shouldn't need to know this.

File: docs/config_example.rst [hunk 12]
```diff
@@ -0,0 +1,131 @@
+Configuration example
+---------------------
+
+Here you can see example objects that can be used for configuration (don't use in a real application)
+
+``'test_auth_provider'`` is dummy example of local auth provider, its linked to the ``'test_identity_provider'`` as specified in ``MULTIPASS_PROVIDER_MAP``.
```
Comment (mvidalgarcia) [inline docs/config_example.rst]: ... is **a** dummy ...
~its~ -> **it's**

File: docs/my_documentation.rst [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+
+``process_login`` then returns the login form of the provider or initializes the external login in case of external provider.
+If no provider is specified the method renders the login selection template specified in ``MULTIPASS_LOGIN_SELECTOR_TEMPLATE``
+
+However, you can also initialize login using the following:
```
Comment (mvidalgarcia) [inline docs/my_documentation.rst]: You forgot something here :smile:

File: docs/my_documentation.rst [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+
+	@multipass.identity_handler
+	def process_identity(identity_info):
+	    identity = Identity.query.filter_by(provider=identity_info.provider.name,
+		                                identifier=identity_info.identifier).first()
+	    if identity is None:
+			user = User.query.filter_by(email=identity_info.data['email']).first()
+			if not user:
+		    	# Redirect to register page
+		    	return redirect(url_for('auth.register', provider=identity_info.provider.name))
```
Comment (mvidalgarcia) [inline docs/my_documentation.rst]: Indentation

File: docs/my_documentation.rst [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+        if next_url:
+            session['_multipass_next_url'] = next_url
+
+If the ``'_multipass_next_url'`` is not set, ``redirect_success`` redirects to ``MULTIPASS_SUCCESS_ENDPOINT`` from Multipass configuration.
+
+
+Failed authentication
+---------------------------
+
+In case the autherntication was unsuccessful, and ``MultipassException`` was raised, ``handle_auth_error`` flashes the ``MULTIPASS_FAILURE_MESSAGE`` and if ``recdirect_to_login`` argument is set, it redirects to ``MULTIPASS_LOGIN_ENDPOINT``
```
Comment (mvidalgarcia) [inline docs/my_documentation.rst]: > In case the authe~r~ntication was ...

Comment (ThiefMaster) [inline docs/my_documentation.rst]: also, `recdirect`?

File: docs/my_documentation.rst [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+
+
+Logout
+---------
+
+``multipass.logout`` should be called by your application upon logout request, passing it the url to redirect to after logout and optionaly a flag to clear session.
```
Comment (mvidalgarcia) [inline docs/my_documentation.rst]: > optional**l**y

File: docs/my_documentation.rst [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+
+.. code-block:: python
+
+	return multipass.logout(request.args.get('next'), clear_session=True)
+
+The ``logout`` method then calls  ``process_logout`` on provider which name was stored in ``session`` as ``'_multipass_login_provider'`` upon login. In ``process_logout``method the provider can implement some provider-specific actions such as sending a logout notification to the provider or redirecting to a SSO logout page. The ``return_url`` from argument can be passed further if the external provider allows to specify the URL to redirect to after logging out.
```
Comment (mvidalgarcia) [inline docs/my_documentation.rst]: The absence of an space after `process_logout` is messing things up in the sentence.

File: docs/my_documentation.rst [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+.. code-block:: python
+
+	criteria['name'] = 'Guinea Pig'
+	criteria['email'] = 'guinea.pig@example.com'
+	results = list(multipass.search_identities(exact=False, **criteria))
+
+Groups
+-------
+
+Providers can divide users into groups. This is usually based on acces rights and competencies of users. For example group of admins, content managers, regular users etc. These groups should be specified in the configuration settings of identity provider. Example from config.example:
```
Comment (mvidalgarcia) [inline docs/my_documentation.rst]: > ... based on acces**s** rights and compete**nc~i~es** ..

File: docs/my_documentation.rst [hunk 19]
```diff
@@ -0,0 +1,365 @@
...
+
+	    supports_member_list = True
+
+	    def get_members(self):
+			members = self.provider.settings['groups'][self.name]
+			for username in members:
+		    	yield self.provider._get_identity(username)
+
+	    def has_member(self, identifier):
+			return identifier in self.provider.settings['groups'][self.name]s
```
Comment (mvidalgarcia) [inline docs/my_documentation.rst]: That `s` at the end. Not sure if it belongs somewhere.

File: docs/my_documentation.rst [hunk 20]
```diff
@@ -0,0 +1,544 @@
...
+.. code-block:: python
+
+	@app.route('/my_login/', methods=('GET', 'POST'))
+	@app.route('/my_login/<provider>', methods=('GET', 'POST'))
+	def login(provider=None):
+	    # Your additional code
+	    return multipass.process_login(provider)
+	
+
+When calling ``process_login`` with provider name as an argument the ``initiate_external_login`` on the provider gets called and redirects user to the provider's site to perform login.( In case of local provider <``_login_form`` link to login of local providers> gets called.) If provider is not specified the ``process_login`` redirects to a login selection template specified in configuration as ``MULTIPASS_LOGIN_SELECTOR_TEMPLATE``.
```
Comment (pferreir) [inline docs/my_documentation.rst]: "with **the** provider name as an argument, the [...]"

Comment (pferreir) [inline docs/my_documentation.rst]: Pay attention to the spacing  between the full stop and opening parenthesis in that line.

Comment (pferreir) [inline docs/my_documentation.rst]: And check the Sphinx documentation on [`:ref:`](http://www.sphinx-doc.org/en/stable/markup/inline.html#role-ref) in order to create a proper link.

File: docs/my_documentation.rst [hunk 21]
```diff
@@ -0,0 +1,544 @@
...
+	    # Your additional code
+	    return multipass.process_login(provider)
+	
+
+When calling ``process_login`` with provider name as an argument the ``initiate_external_login`` on the provider gets called and redirects user to the provider's site to perform login.( In case of local provider <``_login_form`` link to login of local providers> gets called.) If provider is not specified the ``process_login`` redirects to a login selection template specified in configuration as ``MULTIPASS_LOGIN_SELECTOR_TEMPLATE``. 
+
+Identification
+-------------------------
+
+After successful authentication response received from ``initiate_external_login``, method ``handle_auth_success`` gets called.
```
Comment (pferreir) [inline docs/my_documentation.rst]: Again some missing articles:

> After **a** successful authentication response **is** received from ``initiate_external_login``, **the** ``handle_auth_success`` method gets called.

File: docs/my_documentation.rst [hunk 22]
```diff
@@ -0,0 +1,544 @@
...
+
+``handle_auth_success`` then collects the identities linked to the user and stores the name of the provider that was used to login  as ``'_multipass_login_provider'`` in ``session``, it will be used when logging out.
```
Comment (pferreir) [inline docs/my_documentation.rst]: "in ``session``. It will be used when logging out" (note the final stop)

File: docs/my_documentation.rst [hunk 23]
```diff
@@ -0,0 +1,544 @@
...
+
+Then, the method registered via ``'@multipass.identity_handler'`` decorator is called with one, or list of ``IdentityInfo`` objects as an argument. (Depending whether ``MULTIPASS_ALL_MATCHING_IDENTITIES`` is set in configuration).
```
Comment (pferreir) [inline docs/my_documentation.rst]: "via **the** ``'@multipass.identity_handler'`` decorator is called with one, or **a** list of [...]"
"(Depending **on** whether ``MULTIPASS_ALL_MATCHING_IDENTITIES`` is set in **the** configuration)."

File: docs/my_documentation.rst [hunk 24]
```diff
@@ -0,0 +1,544 @@
...
+	@app.route('/my_login/<provider>', methods=('GET', 'POST'))
+	def login(provider=None):
+	    # Your additional code
+	    return multipass.process_login(provider)
+	
+
+When calling ``process_login`` with provider name as an argument the ``initiate_external_login`` on the provider gets called and redirects user to the provider's site to perform login.( In case of local provider <``_login_form`` link to login of local providers> gets called.) If provider is not specified the ``process_login`` redirects to a login selection template specified in configuration as ``MULTIPASS_LOGIN_SELECTOR_TEMPLATE``. 
+
+Identification
+-------------------------
```
Comment (pferreir) [inline docs/my_documentation.rst]: Normally, the underline should have the same number of `-` characters as the title itself.

File: docs/my_documentation.rst [hunk 25]
```diff
@@ -0,0 +1,544 @@
...
+
+After successful authentication response received from ``initiate_external_login``, method ``handle_auth_success`` gets called.
+
+``handle_auth_success`` then collects the identities linked to the user and stores the name of the provider that was used to login  as ``'_multipass_login_provider'`` in ``session``, it will be used when logging out.
+
+Then, the method registered via ``'@multipass.identity_handler'`` decorator is called with one, or list of ``IdentityInfo`` objects as an argument. (Depending whether ``MULTIPASS_ALL_MATCHING_IDENTITIES`` is set in configuration).
+
+Following is our example of ``identity_handler`` method. First, we search through table of identities for one matching ``IdentityInfo.identifier`` from argument. If the identity is not found, we check if we already have this user in the database, we create a new user if needed and assign a new identity to the user object. Then we save this user (or the one found) to the session.
```
Comment (pferreir) [inline docs/my_documentation.rst]: > Here is an example of a possible ``identity_handler`` method: 
> *[put code here]*
> First, we search through **the** table of identities for **a** matching ``IdentityInfo.identifier`` (passed as an argument). If a matching identity is not found, we check if there is already such a user in the database, we create a new user if needed and assign a new identity to that user object. Then, we save the user object (new or matched) in the Flask session.

File: docs/my_documentation.rst [hunk 26]
```diff
@@ -0,0 +1,544 @@
...
+ 
+ 
+ ???? Should I show user and identity classes from example so its more clear how the code works?
```
Comment (pferreir) [inline docs/my_documentation.rst]: I don't think it's needed, as long as we still provide the full example code elsewhere.

**Comment by meluru** [comment]
Good job, I followed your guide and it seemed clear to me

File: docs/quickstart.rst [hunk 27]
```diff
@@ -0,0 +1,553 @@
...
+
+
+Identification
+~~~~~~~~~~~~~~~~
+
+The next step after successful authentication is assigning an identity to the user. That's a job of identity provider so let's have a look how to implement one.
+
+Your identity provider should inherit from ``IdentityProvider`` base class. The most important method it has to implement is ``get_identity_from_auth``. Which accepts ``AuthInfo`` object as an argument and returns the corresponding identity (object of ``IdentityInfo``) based on identifier. 
+
+In our example we search the ``'identities'`` dictionary that we specified in configuration and look for the identity with matching identifier ( 'username' in our case).
```
Comment (mvidalgarcia) [inline docs/quickstart.rst]: `'username'` and the extra space before it.

File: docs/quickstart.rst [hunk 28]
```diff
@@ -0,0 +1,553 @@
...
+.. code-block:: python
+
+	class StaticIdentityProvider(IdentityProvider):
+
+		def get_identity_from_auth(self, auth_info):
+			identifier = auth_info.data['username']	
+			user = self.settings['identities'].get(identifier)
+			if user is None:
+			    return None
+			return IdentityInfo(self, identifier, **user)
```
Comment (mvidalgarcia) [inline docs/quickstart.rst]: I think there's over indentation here as well.

File: docs/quickstart.rst [hunk 29]
```diff
@@ -0,0 +1,553 @@
...
+	criteria['name'] = 'Guinea Pig'
+	criteria['email'] = 'guinea.pig@example.com'
+	results = list(multipass.search_identities(exact=False, **criteria))
+
+.. _groups:
+
+Groups
+-------
+
+Providers can divide users into groups. This is usually based on access rights and competences of users. For example group of admins, content managers, regular users etc. These groups should be specified in the configuration settings of identity provider. Example from our 'test_identity_provider':
```
Comment (mvidalgarcia) [inline docs/quickstart.rst]: `'test_identity_provider'`

File: docs/quickstart.rst [hunk 30]
```diff
@@ -0,0 +1,553 @@
...
+
+If ``multipass.identity_handler`` decorated method doesn't return anything, 
+``handle_auth_success`` returns ``redirect_success`` which redirects to URL stored in ``session`` as ``'_multipass_next_url'``.
+
+.. _external_logout:
+
+External provider logout
+~~~~~~~~~~~~~~~~
+
+``multipass.logout`` should be called by your application upon logout request, passing it the url to redirect to after logout and optionally a flag to clear session.
```
Comment (pferreir) [inline docs/quickstart.rst]: "to clear **the** session".

File: docs/quickstart.rst [hunk 31]
```diff
@@ -0,0 +1,553 @@
...
+
+.. code-block:: python
+
+	@app.route('/logout')
+	def logout():
+	    return multipass.logout(url_for('index'), clear_session=True)
+
+The ``logout`` method then calls  ``process_logout`` on provider which name was stored in ``session`` as ``'_multipass_login_provider'`` upon login.
+
+In ``process_logout`` method the provider can implement some provider-specific actions such as sending a logout notification to the provider or redirecting to a SSO logout page. The ``return_url`` from argument can be passed further if the external provider allows to specify the URL to redirect to after logging out.
```
Comment (pferreir) [inline docs/quickstart.rst]: "In **the** ``process_logout`` method"

File: docs/quickstart.rst [hunk 32]
```diff
@@ -0,0 +1,553 @@
...
+
+Local providers
+********************
+
+Configuration
+~~~~~~~~~~~~~~~~
+
+This section shows an example of a configuration for an application using a local provider. If you wish to use both external and local providers, don't hesitate to specify both local and external providers in the same configuration and just follow our guide also on :ref:`external_providers`
+
+In this example ``'test_auth_provider'`` is a dummy local authentication provider, it's linked to the ``'test_identity_provider'`` as specified in ``MULTIPASS_PROVIDER_MAP``. Specifying ``'type'`` as ``'static'`` links those providers to our ``StaticAuthProvider`` and ``StaticIdentityProvider`` example classes(More on those classes later).
```
Comment (pferreir) [inline docs/quickstart.rst]: Space missing between "classes" and the parenthesis.

File: docs/quickstart.rst [hunk 33]
```diff
@@ -0,0 +1,553 @@
...
+
+In ``identities`` settings of ``'test_auth_provider'`` we specify key-value pairs of username (Pig) and password (pig123), those are used for authentication by Multipass. In this example, the usernames are used as unique identifier for users. In ``identities`` settings of ``'test_identity_provider'`` we assign info keys dictionary to usernames. We also need to specify these keys in ``MULTIPASS_IDENTITY_INFO_KEYS``.
```
Comment (pferreir) [inline docs/quickstart.rst]: I'd say "in the ``identities`` setting of ``test_auth_provider``". I'd also make the username (Pig) and password either italic or code style.

File: docs/quickstart.rst [hunk 34]
```diff
@@ -0,0 +1,553 @@
...
+
+.. code-block:: python
+
+	MULTIPASS_LOGIN_FORM_TEMPLATE = 'login_form.html'
+	
+Implementing providers
+~~~~~~~~~~~~~~~~
+
+Let's create our authentication provider class, which should inherit from ``AuthProvider``.
+We should also specify the login form class (which inherits from ``FlaskForm``) which we use for login of this provider.
```
Comment (pferreir) [inline docs/quickstart.rst]: "for **the** login operation of this provider"

File: docs/quickstart.rst [hunk 35]
```diff
@@ -0,0 +1,553 @@
...
+
+	class StaticAuthProvider(AuthProvider):
+	    login_form = StaticLoginForm
+    
+.. _local_login:
+
+Local provider login
+~~~~~~~~~~~~~~~~
+
+The process of handling URLs for login is the same as with external providers therefore please check the :ref:`external_login` part of this guide.
```
Comment (pferreir) [inline docs/quickstart.rst]: Missing comma between "providers" and "therefore".

File: docs/quickstart.rst [hunk 36]
```diff
@@ -0,0 +1,553 @@
...
+
+The only difference is that  ``multipass.process_login`` calls method ``_login_form`` which renders a template specified in ``MULTIPASS_LOGIN_FORM_TEMPLATE`` with the ``login_form`` specified in the authentication provider class.
```
Comment (pferreir) [inline docs/quickstart.rst]: "calls **the** method".

File: docs/quickstart.rst [hunk 37]
```diff
@@ -0,0 +1,553 @@
...
+
+Once the form is submitted method ``process_local_login`` of authentication provider class is called. In this method you have to implement your authentication logic.
```
Comment (pferreir) [inline docs/quickstart.rst]: "Once the form is submitted *comma* **the** method... of **the** authentication provider class..."

File: docs/quickstart.rst [hunk 38]
```diff
@@ -0,0 +1,553 @@
...
+Local provider login
+~~~~~~~~~~~~~~~~
+
+The process of handling URLs for login is the same as with external providers therefore please check the :ref:`external_login` part of this guide.
+
+The only difference is that  ``multipass.process_login`` calls method ``_login_form`` which renders a template specified in ``MULTIPASS_LOGIN_FORM_TEMPLATE`` with the ``login_form`` specified in the authentication provider class.
+
+Once the form is submitted method ``process_local_login`` of authentication provider class is called. In this method you have to implement your authentication logic. 
+
+You should raise ``MultipassException`` in case of failed validation. If the validation was successful, ``AuthInfo`` object should be created and passed to ``multipass.handle_auth_success``. Bellow is ``process_local_login`` method from our example provider ``StaticAuthProvider``:
```
Comment (pferreir) [inline docs/quickstart.rst]: **the** ``AuthInfo`` object

Comment (pferreir) [inline docs/quickstart.rst]: "Below" has only one "l".

Comment (pferreir) [inline docs/quickstart.rst]: "Below is **the** ``process_local_login``..."

File: docs/quickstart.rst [hunk 39]
```diff
@@ -0,0 +1,553 @@
...
+	    if password != data['password']:
+	        raise AuthenticationFailed('Invalid password.')
+	    auth_info = AuthInfo(self, username=data['username'])
+	    return self.multipass.handle_auth_success(auth_info)
+
+
+Identification
+~~~~~~~~~~~~~~~~
+
+The next step after successful authentication is assigning an identity to the user. That's a job of identity provider so let's have a look how to implement one.
```
Comment (pferreir) [inline docs/quickstart.rst]: "That's a job for **an** identity provider, so..."

File: docs/quickstart.rst [hunk 40]
```diff
@@ -0,0 +1,553 @@
...
+
+Your identity provider should inherit from ``IdentityProvider`` base class. The most important method it has to implement is ``get_identity_from_auth``. Which accepts ``AuthInfo`` object as an argument and returns the corresponding identity (object of ``IdentityInfo``) based on identifier.
```
Comment (pferreir) [inline docs/quickstart.rst]: "from **the**``IdentityProvider`` base class"

Comment (pferreir) [inline docs/quickstart.rst]: There should be a comma before "Which", not a full stop.
"based on **an** identifier"

File: docs/quickstart.rst [hunk 41]
```diff
@@ -0,0 +1,553 @@
...
+
+
+Identification
+~~~~~~~~~~~~~~~~
+
+The next step after successful authentication is assigning an identity to the user. That's a job of identity provider so let's have a look how to implement one.
+
+Your identity provider should inherit from ``IdentityProvider`` base class. The most important method it has to implement is ``get_identity_from_auth``. Which accepts ``AuthInfo`` object as an argument and returns the corresponding identity (object of ``IdentityInfo``) based on identifier. 
+
+In our example we search the ``'identities'`` dictionary that we specified in configuration and look for the identity with matching identifier (``'username'`` in our case).
```
Comment (pferreir) [inline docs/quickstart.rst]: "with **a** matching identifier"

File: docs/quickstart.rst [hunk 42]
```diff
@@ -0,0 +1,553 @@
...
+		identifier = auth_info.data['username']	
+		user = self.settings['identities'].get(identifier)
+		if user is None:
+		    return None
+		return IdentityInfo(self, identifier, **user)
+			
+Other methods that should be implemented to ensure the full Multipass functionality can be found further in this guide. See :ref:`identities` and :ref:`groups`
+
+Now let's get back to the identification process. 
+Once ``handle_auth_success`` is called, it collects the identities linked to the user using the ``get_identity_from_auth`` we just mentioned.
```
Comment (pferreir) [inline docs/quickstart.rst]: "using the ``get_identity_from_auth`` **method** we just mentioned"

File: docs/quickstart.rst [hunk 43]
```diff
@@ -0,0 +1,553 @@
...
+		user = self.settings['identities'].get(identifier)
+		if user is None:
+		    return None
+		return IdentityInfo(self, identifier, **user)
+			
+Other methods that should be implemented to ensure the full Multipass functionality can be found further in this guide. See :ref:`identities` and :ref:`groups`
+
+Now let's get back to the identification process. 
+Once ``handle_auth_success`` is called, it collects the identities linked to the user using the ``get_identity_from_auth`` we just mentioned. 
+Once identities are successfully collected, the method registered via ``'@multipass.identity_handler'`` decorator is called. A method with this decorator must be implemented in your application. Check the :ref:`identity_handler` part of this documentation for more info.
```
Comment (pferreir) [inline docs/quickstart.rst]: "via **the** ``'@multipass.identity_handler'`` decorator"

File: docs/quickstart.rst [hunk 44]
```diff
@@ -0,0 +1,553 @@
...
+
+Failed authentication
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In case the authentication was unsuccessful, and ``MultipassException`` was raised, ``handle_auth_error`` flashes the ``MULTIPASS_FAILURE_MESSAGE`` and if ``redirect_to_login`` argument is set, it redirects to ``MULTIPASS_LOGIN_ENDPOINT``
```
Comment (pferreir) [inline docs/quickstart.rst]: "if **the** ``redirect_to_login`` argument"

File: docs/quickstart.rst [hunk 45]
```diff
@@ -0,0 +1,553 @@
...
+
+
+Local provider logout
+~~~~~~~~~~~~~~~~
+
+Process of logging out local providers is the same as with external providers.
```
Comment (pferreir) [inline docs/quickstart.rst]: "**The** process of logging out local providers"

File: docs/quickstart.rst [hunk 46]
```diff
@@ -0,0 +1,553 @@
...
+Please check the :ref:`external_logout` part of this guide.
+
+.. _identities:
+
+Identities
+----------
+
+To retrieve ``IdentityInfo`` object  your  ``IdentityProvider`` must implement ``get_identity`` method. Example from ``StaticIdentityProvider``:
```
Comment (pferreir) [inline docs/quickstart.rst]: "To retrieve **an** ``IdentityInfo`` object *comma*  your  ``IdentityProvider`` must implement **the** ``get_identity`` method. Example from ``StaticIdentityProvider``"

File: docs/quickstart.rst [hunk 47]
```diff
@@ -0,0 +1,553 @@
...
+
+.. code-block:: python
+
+    def get_identity(self, identifier):
+        user = self.settings['identities'].get(identifier)
+        if user is None:
+            return None
+        return IdentityInfo(self, identifier, **user)
+
+Same applies for searching identities. There you accept ``criteria`` dictionary as a filter for your search. Example from ``StaticIdentityProvider``:
```
Comment (pferreir) [inline docs/quickstart.rst]: "**The** same [...] There you should expect a ``criteria`` dictionary..."

File: docs/quickstart.rst [hunk 48]
```diff
@@ -0,0 +1,553 @@
...
+                if not any(user_values):
+                    break
+                elif exact and not user_values & set(values):
+                    break
+                elif not exact and not any(sv in uv for sv, uv in itertools.product(values, user_values)):
+                    break
+            else:
+                yield IdentityInfo(self, identifier, **user)
+
+Once implemented on your  ``IdentityProvider`` you can also use method  ``search_identities`` on ``Multipass`` object which will search among all the providers and yield all the ``identity_info``  matching the criteria specified in the argument.
```
Comment (pferreir) [inline docs/quickstart.rst]: "Once implemented on your  ``IdentityProvider`` *comma* you "

Comment (pferreir) [inline docs/quickstart.rst]: "on **the/a** ``Multipass`` object"

File: docs/quickstart.rst [hunk 49]
```diff
@@ -0,0 +1,553 @@
...
+	criteria['name'] = 'Guinea Pig'
+	criteria['email'] = 'guinea.pig@example.com'
+	results = list(multipass.search_identities(exact=False, **criteria))
+
+.. _groups:
+
+Groups
+-------
+
+Providers can divide users into groups. This is usually based on access rights and competences of users. For example group of admins, content managers, regular users etc. These groups should be specified in the configuration settings of identity provider. Example from our ``'test_identity_provider'``:
```
Comment (pferreir) [inline docs/quickstart.rst]: "[...] on **the** access rights and competences of users, for example: whether they are admins, content managers, regular users, etc."
"configuration settings of **the** identity provider"

File: docs/quickstart.rst [hunk 50]
```diff
@@ -0,0 +1,553 @@
...
+		    'Bunny': {'email': 'bugs.bunny@example.com', 'name': 'Bugs Bunny'}
+		},
+		'groups': {
+		    'Admins': ['Pig'],
+		    'Everybody': ['Pig', 'Bunny'],
+		}
+	    }
+
+
+Provider's group class must inherit from base class ``Group``. If the group should support members, methods ``get_members`` (returning iterable of ``IdentityInfo`` of the group members) and ``has_member`` must be implemented. Example from our ``StaticGroup``:
```
Comment (pferreir) [inline docs/quickstart.rst]: "**The** provider's group class must inherit from **the** base class ``Group``"

File: docs/quickstart.rst [hunk 51]
```diff
@@ -0,0 +1,553 @@
...
+
+	    def get_members(self):
+		members = self.provider.settings['groups'][self.name]
+		for username in members:
+	    		yield self.provider._get_identity(username)
+
+	    def has_member(self, identifier):
+		return identifier in self.provider.settings['groups'][self.name]
+		
+In your ``IdentityProvider`` class you must specify the group class as ``group_class`` and flag ``supports_groups`` must be set.
```
Comment (pferreir) [inline docs/quickstart.rst]: "and **the** flag"

File: docs/quickstart.rst [hunk 52]
```diff
@@ -0,0 +1,553 @@
...
+	MULTIPASS_IDENTITY_INFO_KEYS = ['email', 'name', 'affiliation']
+
+We also need to specify the template with a login form for our provider:
+
+.. code-block:: python
+
+	MULTIPASS_LOGIN_FORM_TEMPLATE = 'login_form.html'
+	
+Implementing providers
+~~~~~~~~~~~~~~~~
```
Comment (ThiefMaster) [inline docs/quickstart.rst]: many of these underlines don't match the length of the line they are associated with

**Comment by ThiefMaster** [comment]
Let's get rid of the git submodule for the themes and instead install it from pip (like we do e.g. in the flask-pluginengine docs).

https://github.com/pallets/pallets-sphinx-themes

**Comment by ThiefMaster** [comment]
Use the Flask theme and add its package to all-requirements.txt; once that's done I think it can finally be merged.
